cpp/src/arrow/compute/expression_internal.h

```diff
@@ -278,8 +278,11 @@ struct FlattenedAssociativeChain {
 
 inline Result<std::shared_ptr<compute::Function>> GetFunction(
     const Expression::Call& call, compute::ExecContext* exec_context) {
-  if (call.function_name != "cast") {
-    return exec_context->func_registry()->GetFunction(call.function_name);
+  auto input_call_function =
+      exec_context->func_registry()->GetFunction(call.function_name);
+  auto cast_function = exec_context->func_registry()->GetFunction("cast");
+  if (input_call_function != cast_function) {
+    return input_call_function;
```

Comment:
This is not ideal since it doubles the overhead of GetFunction. Instead, let's add a lazily initialized `Function* FunctionRegistry::cast_function() const` property so that this can be rewritten
```suggestion
  auto function = exec_context->func_registry()->GetFunction(call.function_name);
  if (function != exec_context->func_registry()->cast_function()) {
    return function;
```

Reply:
Actually instead of lazy initialization inside `cast_function()`, it'd be more efficient to initialize it inside `AddFunction()` when the function named "cast" is initialized

Reply:
Thanks for your suggestion, it's much better than before.